packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -29,64 +38,157 @@ class UpgradeCommand extends FlutterCommand {
 
   @override
   Future<FlutterCommandResult> runCommand() async {
+    final UpgradeCommandRunner upgradeCommandRunner = UpgradeCommandRunner();
+    await upgradeCommandRunner.runCommand(argResults['force'], GitTagVersion.determine(), FlutterVersion.instance);
+    return null;
+  }
+}
+
+
+@visibleForTesting
+class UpgradeCommandRunner {
+  Future<FlutterCommandResult> runCommand(bool force, GitTagVersion gitTagVersion, FlutterVersion flutterVersion) async {
+    await verifyUpstreamConfigured();
+    if (!force && gitTagVersion == const GitTagVersion.unknown()) {
+      // If the commit is a recognized branch and not master,
+      // explain that we are avoiding potential damage.
+      if (flutterVersion.channel != 'master' && FlutterVersion.officialChannels.contains(flutterVersion.channel)) {
+        throwToolExit(
+          'Unknown flutter tag. Abandoning upgrade to avoid destroying local '
+          'changes. It is recommended to use git directly if not working off of '
+          'an official channel.'
+        );
+      // Otherwise explain that local changes can be lost.
+      } else {
+        throwToolExit(
+          'Unknown flutter tag. Abandoning upgrade to avoid destroying local '
+          'changes. If it is okay to remove local changes, then re-run this '
+          'command with --force.'
+        );
+      }
+    }
+    final bool stashedChanges = await maybeStashChanges(gitTagVersion);
+    await updateChannel(flutterVersion);
+    await attemptRebase();
+    await updateEngine();
+    await updatePackages(flutterVersion);
+    await runDoctor();
+    if (stashedChanges) {
+      await popChanges();
+    }
+    return null;
+  }
+
+  /// Check if there is an upstream repository configured.
+  ///
+  /// Exits tool if there is no upstream.
+  Future<void> verifyUpstreamConfigured() async {
     try {
       await runCheckedAsync(<String>[
         'git', 'rev-parse', '@{u}',
       ], workingDirectory: Cache.flutterRoot);
     } catch (e) {
       throwToolExit('Unable to upgrade Flutter: no upstream repository configured.');
     }
+  }
 
-    final FlutterVersion flutterVersion = FlutterVersion.instance;
+  /// Attempt to stash any local changes, returning true if successufl.
+  ///
+  /// Exits tool if `git stash` returns a non-zero exit code.
+  Future<bool> maybeStashChanges(GitTagVersion gitTagVersion) async {
+    bool stashedChanges = false;
+    final String stashName = 'flutter-upgrade-from-v${gitTagVersion.x}.${gitTagVersion.y}.${gitTagVersion.z}';
+    try {
+      final RunResult runResult = await runCheckedAsync(<String>[
+        'git', 'stash', 'push', '-m', stashName
+      ]);
+      // output message will contain stashname if successful.
+      if (runResult.stdout.contains(stashName)) {
+        stashedChanges = true;
+      }
+    } catch (e) {
+      throwToolExit('Failed to stash local changes: $e');
+    }
+    return stashedChanges;
+  }
 
+  /// Attempts to upgrade the channel.
```

Comment:
I'd forgotten about channel upgrade.  Can you update this to say that it attempts to migrate the user off a deprecated channel if needed?